File: backend/src/crypto.js
import crypto from 'crypto';

/**
 * Encrypts a buffer using AES-256-GCM.
 * The 16-byte authentication tag is appended to the end of the encrypted buffer.
 * @param {Buffer} buffer - Content to encrypt
 * @param {string} keyHex - 64-character hexadecimal key (256 bits)
 * @returns {{encrypted: Buffer, ivHex: string}}
 */
export const encryptBuffer = (buffer, keyHex) => {
  const key = Buffer.from(keyHex, 'hex');
  if (key.length !== 32) {
    throw new Error('Invalid key length. Key must be a 32-byte (256-bit) hex string.');
  }

  // 12-byte IV is standard for GCM
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  
  const encrypted = Buffer.concat([
    cipher.update(buffer),
    cipher.final()
  ]);
  
  const authTag = cipher.getAuthTag(); // 16 bytes
  
  // Combine encrypted data and authTag into a single buffer
  const combined = Buffer.concat([encrypted, authTag]);
  
  return {
    encrypted: combined,
    ivHex: iv.toString('hex')
  };
};

/**
 * Decrypts a buffer using AES-256-GCM.
 * The authentication tag is extracted from the last 16 bytes of the buffer.
 * @param {Buffer} encryptedBufferCombined - Ciphertext with appended 16-byte auth tag
 * @param {string} keyHex - 64-character hexadecimal key (256 bits)
 * @param {string} ivHex - Hexadecimal initialization vector
 * @returns {Buffer} decrypted buffer
 */
export const decryptBuffer = (encryptedBufferCombined, keyHex, ivHex) => {
  const key = Buffer.from(keyHex, 'hex');
  const iv = Buffer.from(ivHex, 'hex');

  if (key.length !== 32) {
    throw new Error('Invalid key length. Key must be a 32-byte (256-bit) hex string.');
  }

  // Extract auth tag from the end of the buffer (last 16 bytes)
  const authTagLength = 16;
  const encryptedDataLength = encryptedBufferCombined.length - authTagLength;
  
  const encryptedData = encryptedBufferCombined.subarray(0, encryptedDataLength);
  const authTag = encryptedBufferCombined.subarray(encryptedDataLength);

  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
  decipher.setAuthTag(authTag);

  return Buffer.concat([
    decipher.update(encryptedData),
    decipher.final()
  ]);
};

export default {
  encryptBuffer,
  decryptBuffer
};

File: backend/src/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"io"
)

// EncryptBuffer encrypts cleartext using AES-256-GCM.
// The 16-byte authentication tag is appended to the end of the returned ciphertext slice.
func EncryptBuffer(plaintext []byte, keyHex string) ([]byte, string, error) {
	key, err := hex.DecodeString(keyHex)
	if err != nil {
		return nil, "", fmt.Errorf("failed to decode key hex: %w", err)
	}
	if len(key) != 32 {
		return nil, "", fmt.Errorf("invalid key size: key must be 32 bytes (256 bits)")
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, "", fmt.Errorf("failed to create cipher block: %w", err)
	}

	// 12-byte IV is standard for GCM
	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, "", fmt.Errorf("failed to create GCM block: %w", err)
	}

	iv := make([]byte, 12)
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, "", fmt.Errorf("failed to generate random IV: %w", err)
	}

	// Seal appends the 16-byte tag to the ciphertext buffer automatically
	ciphertext := aesgcm.Seal(nil, iv, plaintext, nil)

	return ciphertext, hex.EncodeToString(iv), nil
}

// DecryptBuffer decrypts ciphertext (with appended 16-byte tag) using AES-256-GCM.
func DecryptBuffer(ciphertext []byte, keyHex string, ivHex string) ([]byte, error) {
	key, err := hex.DecodeString(keyHex)
	if err != nil {
		return nil, fmt.Errorf("failed to decode key hex: %w", err)
	}
	if len(key) != 32 {
		return nil, fmt.Errorf("invalid key size: key must be 32 bytes (256 bits)")
	}

	iv, err := hex.DecodeString(ivHex)
	if err != nil {
		return nil, fmt.Errorf("failed to decode IV hex: %w", err)
	}
	if len(iv) != 12 {
		return nil, fmt.Errorf("invalid IV size: IV must be 12 bytes")
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("failed to create cipher block: %w", err)
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("failed to create GCM block: %w", err)
	}

	// Open decrypts ciphertext and verifies the appended 16-byte tag
	plaintext, err := aesgcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decryption failed: %w", err)
	}

	return plaintext, nil
}
